### DISCUSSION TITLE
watchAsset method

### DISCUSSION THREAD

**Discussion by mrfujisawa** [wevm/wagmi#639] [Q&A] [answered]
What is the intended pattern for adding an asset to the user's wallet?  This doesn't seem to be documented anywhere in WAGMI (maybe i missed it) but if I wanted to add an ERC20 to a user's wallet this is the recommended way to do so?

```
// pseudo-ish code
const { activeConnector } = useConnect();
const wasAdded = await activeConnector!.watchAsset!(erc20Asset);
```
essentially the user needs to be connected to add an asset (needs an activeConnector) and the connector needs to have the `watchAsset` method defined.  Am i doing this right or is there a better way to do this?

**Comment by tmm** [comment, marked as answer]
Yup! That will work. Version `0.2.x` had this built in as part of `useToken`, but never made the jump to later versions.

We could add a `useWatchAsset` hook that follows a similar implementation to `useSwitchNetwork` (since that isn't supported across all connectors either).

**Reply by ECBSJ** [reply]
> Yup! That will work. Version `0.2.x` had this built in as part of `useToken`, but never made the jump to later versions.
> 
> We could add a `useWatchAsset` hook that follows a similar implementation to `useSwitchNetwork` (since that isn't supported across all connectors either).

What's the current plan on adding a `useWatchAsset` hook?

And in regards to the pseudo-ish code in the initial question, is `watchAsset` already a defined method on `activeConnector`? I'm having troubles finding it as a method.

**Comment by einaralex** [comment]
Since `useToken` is now marked as deprecated, I implemented my own solution.

```js
import { EIP1193Provider } from 'viem';
import { useAccount } from 'wagmi';

export const useWatchAsset = () => {
  const { isConnected, connector } = useAccount();

  if (!isConnected || !connector) {
    return { watchAsset: null };
  }

  return {
    watchAsset: async (token) => {
      try {
        const provider = await connector.getProvider();

        await (provider as EIP1193Provider)?.request({
          // See details: https://docs.metamask.io/wallet/reference/wallet_watchasset/
          method: 'wallet_watchAsset',
          params: {
            type: 'ERC20',
            options: token,
          },
        });
      } catch (error) {
        console.error('Failed to watch asset:', error);
      }
    },
  };
};
```

Usage
```js
const { watchAsset } = useWatchAsset();


if (watchAsset) {
 // For Metamask
  watchAsset({ address: tokenAddress })
  
  // For Rabby
  watchAsset({ address: tokenAddress, chain: "ethereum" })
}
```
